File: 2019/Problem J3: Cold Compress.go
package main

import (
  "fmt"
  "strconv"
  "strings"
)

func output(str string) []string{
  index := 0
  var index2 int
  var portion string
  out_list := make([]string, 0, 0)
  str += " "

  for true {
    count := 0

    for i := index; i < len(str); i++ {
      if str[i] != str[index] {
        index2 = i-1
        break
      }
      count ++
    }

    if index2 - index > 0 {
      portion = str[index:index2]
    } else {
      portion = string(str[index])
      index2 ++
    } 

    out_list = append(out_list, strconv.Itoa(count), string(portion[0]))

    index += count

    if index+1 == len(str) {
      break
    }
  }
  return out_list
}

func main() {
  var n int
  lines := make([]string, 0, 0)
  fmt.Scan(&n)
  for i := 0; i < n; i++  {
    var input string
    fmt.Scan(&input)
    lines = append(lines, input)
  }
  for _, i := range lines {
    fmt.Println(strings.Join(output(i), " "))
  }
}

File: archive/2019/Problem J3: Cold Compress.cpp
#include <iostream>
#include <vector>

using std::cout;
using std::cin;
using std::endl;
using std::string;

std::vector<string> output(string str) {
    int index = 0;
    int index2;
    int count;
    string portion;
    std::vector<string> out_list;
    str.append(" ");

    while (true) {
        count = 0;
        for (unsigned int i = index; i < str.length(); i++) {
            if (str[i] != str[index]) {
                index2 = i-1;
                break;
            }
            count++;
        }
        if (index2 - index > 0) {
            portion = str.substr(index, index2-index);
        } else {
            portion = str[index];
            index2++;
        }
        out_list.push_back(std::to_string(count));
        string out;
        out += portion[0];
        out_list.push_back(out);

        index += count;

        if (index+1 == str.length()) {
            break;
        }
    }
    return out_list;
}

int main() {
    string n;
    getline(cin, n);
    cin.clear();
    string lines[std::stoi(n)];

    for (unsigned int i = 0; i < std::stoi(n); i++) {
        string a;
        getline(cin, a);
        lines[i] = a;
    }

    for (const string &i: lines) {
        std::vector<string> res = output(i);
        for (const string &x: res) {
            cout << x << " ";
        }
        cout << endl;
    }

    return 0;
}
